### PR TITLE
Adding React and TypeScript

### PR DESCRIPTION
**Pull request by emilkrebs**
Everything inside the preview window was rendered with [React](https://github.com/langium/langium-website/blob/d6e67ea0c26d80bd7eb60bc7710b7ab3c43bec68/hugo/assets/scripts/statemachine.tsx)

Just link the file with the "file" parameter and you are ready to go. 
For example: `file: "scripts/statemachine.tsx"`

Edit: 
- recreated the style with Tailwind V3 and made it responsive
- removed strict mode

preview

### CODE REVIEW

File: hugo/content/showcase/statemachine.html [hunk 1]
```diff
@@ -5,12 +5,38 @@
 layout: showcase-page
 url: "/showcase/statemachine"
 img: "/assets/Langium_Statemachine.svg"
+file: "scripts/statemachine.tsx"
```
Comment (emilkrebs) [inline hugo/content/showcase/statemachine.html:8]: Here the magic goes

File: hugo/babel.config.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module.exports = function (api) {
+    api.cache(true);
+    const presets = [
+      ["@babel/preset-react"],
+      ["@babel/preset-typescript"]
```
Comment (emilkrebs) [inline hugo/babel.config.js:5]: You need this to pipe TypeScript through babel.

File: hugo/layouts/langium/showcase-page.html [hunk 3]
```diff
@@ -1,24 +1,65 @@
...
+    <div class="preview-window" id="preview">
+    </div>
 </div>
+
+
+
 {{ .Content }}
+
+{{ if .Params.file }}
+    {{- $jsFile := resources.Get .Params.file | resources.ExecuteAsTemplate "page.js" . | babel | js.Build -}}
```
Comment (emilkrebs) [inline hugo/layouts/langium/showcase-page.html:61]: This line pipes the specified react file through babel.

File: hugo/layouts/langium/showcase-page.html [hunk 4]
```diff
@@ -1,24 +1,65 @@
...
+    <script type="module" src="{{ $jsFile.RelPermalink }}"></script>
```
Comment (emilkrebs) [inline hugo/layouts/langium/showcase-page.html:62]: This line adds the code to the page.

File: hugo/assets/scripts/statemachine.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+     <div className='placeholder'>
+      Preview not available
+     </div>
+    </div>
+  );
+}
+
+const root = createRoot(document.getElementById("preview") as HTMLElement);
+root.render(
+  <React.StrictMode>
```
Comment (Lotes) [inline hugo/assets/scripts/statemachine.tsx:16]: afaik the strict mode renders the inner component twice. Within prod code you should remove it. Maybe read about it and then decide.

File: hugo/content/showcase/statemachine.html [hunk 6]
```diff
@@ -5,12 +5,38 @@
...
 description: A language that captures the functionality of a state machine. Demonstrated by modeling a traffic light.
 geekdochidden: true
 draft: false
 noMain: true
 ---
 
+<style>
```
Comment (Lotes) [inline hugo/content/showcase/statemachine.html:15]: Please use Tailwind css classes instead

File: hugo/assets/scripts/statemachine.tsx (hunk 5 above)
Comment (montymxb) [reply hugo/assets/scripts/statemachine.tsx:16]: According to [React's docs about strict mode](https://reactjs.org/docs/strict-mode.html), this shouldn't run in production, should be okay

File: hugo/content/showcase/statemachine.html (hunk 6 above)
Comment (montymxb) [reply hugo/content/showcase/statemachine.html:15]: I'm inclined to agree as well. When possible using the [tailwindcss utility classes](https://tailwindcss.com/docs/utility-first) that are provided is also a pretty good bet. It makes the styling very portable (i.e. the classes can go anywhere and still achieve the same effect). However if you run into a few special cases where you can't quite find the correct utility class, I would recommend updating the stylesheet in the tailwinds folder.

Although I have also definitely written inline stylesheets for other parts of the showcase, but I probably should have been using tailwind as well in those contexts.

File: hugo/layouts/langium/showcase-page.html [hunk 7]
```diff
@@ -7,18 +7,58 @@
         height: 100%;
     }
 
+    .container {
+        width: 100%;
+        height: 100%;
+        border: 1px rgba(38, 136, 140) solid;
+        justify-content: center;
+        align-self: center;
+        display: flex;
```
Comment (montymxb) [inline hugo/layouts/langium/showcase-page.html:16]: There's an issue with layout at pages that have a smaller height than expected, where monaco greatly exceeds the container bounds. (was using responsive design mode in Firefox, and pushed the page to a compact 2000x500)
wide
You can hide the overflow like so, and monaco looks like it compensates just fine.
```suggestion
        display: flex;
        overflow: hidden;
```

File: hugo/layouts/langium/showcase-page.html [hunk 8]
```diff
@@ -7,18 +7,58 @@
...
+    }
+
+    .editor {
```
Comment (montymxb) [inline hugo/layouts/langium/showcase-page.html:19]: Just a thought, on smaller screen sizes, like a phone, it's very difficult to fit both the editor & the preview side-by-side. I would recommend either using a media-query to break down these elements into one on top of the other, or to add some sort of 'tab' mechanism at the top of the page that allows switching between the two displays on smaller screens. The first approach would work and would be quicker, the second would take more time, and is maybe something we can consider later on.

Comment (montymxb) [reply hugo/layouts/langium/showcase-page.html:19]: Here's an example.
squish

**Review by montymxb — COMMENTED** [review]
This looks really nice! Thanks for the work on this.

Some general suggestions that I would add. In addition, I was also wondering, while we just have a 'mock preview' this would make sense to keep this in a separate branch until it's ready to add. It would be nice to only update the live site when the preview functionality is complete (or at least functional) for the statemachine.

File: tailwind/style.css [hunk 9]
```diff
@@ -61,5 +61,5 @@
 }
 
 .feature-item-content {
-    @apply h-full w-full box-border rounded-xl border-8 border-emeraldLangium dark:border-emeraldLangium sm:px-6 px-2 pb-6 mx-2 sm:mx-0 bg-emeraldLangium dark:bg-gray-900 shadow-lg;
+    @apply h-full w-full box-border rounded-xl border-2 border-emeraldLangium dark:border-emeraldLangium sm:px-6 px-2 pb-6 mx-2 sm:mx-0 bg-emeraldLangium dark:bg-gray-900 shadow-lg;
```
Comment (emilkrebs) [inline tailwind/style.css:64]: Oh, yes, I made the border smaller. Don't know if it really looks better but I think so.

File: hugo/assets/scripts/statemachine.tsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { createRoot } from 'react-dom/client';
+
+function Preview() {
+  return (
+    <div className="flex h-full w-full items-center justify-center">
+     <div className="text-white border-2 border-solid border-emeraldLangium rounded-md p-4 w-8 text-center text-sm shadow-opacity-50 hover:shadow-[0px_0px_25px_0px] hover:shadow-emeraldLangiumDarker">
```
Comment (emilkrebs) [inline hugo/assets/scripts/statemachine.tsx:7]: Here goes tailwind

File: hugo/assets/scripts/statemachine.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+      Preview not available
+     </div>
+    </div>
+  );
+}
+
+const root = createRoot(document.getElementById("preview") as HTMLElement);
+root.render(
+    <Preview />
```
Comment (emilkrebs) [inline hugo/assets/scripts/statemachine.tsx:16]: No more strict mode

File: hugo/content/showcase/statemachine.html [hunk 12]
```diff
@@ -101,5 +102,8 @@
 
     client.startEditor(document.getElementById("monaco-editor-root"));
 
-    window.addEventListener("resize", () => client.updateLayout());
+    window.addEventListener("resize", () => {
+        client.updateLayout({width: 100, heigth: 100})
```
Comment (emilkrebs) [inline hugo/content/showcase/statemachine.html:106]: Fixed weird resize behavior of the Monaco Editor

File: hugo/layouts/langium/showcase-page.html [hunk 13]
```diff
@@ -7,18 +7,35 @@
...
+        width: 100%;
     }
 
     #monaco-editor-root {
         flex-grow: 1;
     }
 </style>
-<div class="wrapper relative bg-white dark:bg-gray-900">
-    <div class="dark:bg-gray-900" id="monaco-editor-root"></div>
+<div class="w-full h-full border border-emeraldLangium justify-center self-center flex">
```
Comment (emilkrebs) [inline hugo/layouts/langium/showcase-page.html:22]: More tailwind...

File: package.json [hunk 14]
```diff
@@ -11,12 +11,17 @@
...
+    "@types/react": "^18.0.21",
+    "@types/react-dom": "^18.0.6",
+    "autoprefixer": "^10.4.12",
     "concurrently": "^6.2.2",
     "cross-env": "^7.0.3",
     "hugo-extended": "^0.89.4",
+    "postcss": "^8.4.18",
     "shx": "^0.3.4",
-    "tailwindcss": "^2.2.16"
+    "tailwindcss": "^3.1.8"
```
Comment (emilkrebs) [inline package.json:24]: Updated tailwind so I can use shadows

File: tailwind/tailwind.config.js [hunk 15]
```diff
@@ -1,8 +1,7 @@
 const colors = require('tailwindcss/colors')
 
 module.exports = {
-  mode: 'jit',
-  purge: ['../hugo/layouts/**/*.html','../hugo/content/**/*.html','../hugo/static/**/*.html','../hugo/static/**/*.js'],
+  content: ['../hugo/layouts/**/*.html','../hugo/content/**/*.html','../hugo/static/**/*.html','../hugo/static/**/*.js', '../hugo/assets/scripts/*.tsx', '../hugo/assets/scripts/*.jsx'],
```
Comment (emilkrebs) [inline tailwind/tailwind.config.js:4]: Added react files to tailwind.

**Review by montymxb — APPROVED** [review]
Approved :octocat:! This gets a really convenient skeleton in place to work with. Also very happy putting this into a preview branch, so we can continue to hack away until this is all ready to go.

There's still a couple little things with the styling, but that can be resolved later, so the next steps can proceed 👍

**Pull request merged**